### PR TITLE
Introduce Auth propagate policy from sources like request headers and query params

### PR DESCRIPTION
**Pull request by neelayu**
**Description:** 
Enhancement: ServerAuth now defines an additional field `propagate` to define what needs to be passed on to the interceptor.

**Link to tracking Issue:** #4806

**Testing:**  Unit tests added and manual testing

@jpkrohling

### CODE REVIEW

File: config/configauth/serverauth.go [hunk 1]
```diff
@@ -28,17 +28,18 @@ import (
 type ServerAuthenticator interface {
 	component.Extension
 
-	// Authenticate checks whether the given headers map contains valid auth data. Successfully authenticated calls will always return a nil error.
+	// Authenticate checks whether the given requestMap contains valid auth data. requestMap consists of request headers and query parameters as specified in the PropagatePolicy.
+	// For gRPC requests, request map includes headers from the metadata. Successfully authenticated calls will always return a nil error.
```
Comment (jpkrohling) [inline config/configauth/serverauth.go:32]:
```suggestion
	// For gRPC requests, "metadata" includes headers from the RPC's metadata. Successfully authenticated calls will always return a nil error.
```

File: config/configauth/serverauth.go [hunk 2]
```diff
@@ -28,17 +28,18 @@ import (
 type ServerAuthenticator interface {
 	component.Extension
 
-	// Authenticate checks whether the given headers map contains valid auth data. Successfully authenticated calls will always return a nil error.
+	// Authenticate checks whether the given requestMap contains valid auth data. requestMap consists of request headers and query parameters as specified in the PropagatePolicy.
```
Comment (jpkrohling) [inline config/configauth/serverauth.go:31]:
```suggestion
	// Authenticate checks whether the given "metadata" contains valid auth data. For HTTP requests, the parameter "metadata" consists of request headers and query parameters as specified in the PropagatePolicy.
```

File: config/configauth/serverauth.go [hunk 3]
```diff
@@ -28,17 +28,18 @@ import (
...
 	// When the authentication fails, an error must be returned and the caller must not retry. This function is typically called from interceptors,
 	// on behalf of receivers, but receivers can still call this directly if the usage of interceptors isn't suitable.
 	// The deadline and cancellation given to this function must be respected, but note that authentication data has to be part of the map, not context.
 	// The resulting context should contain the authentication data, such as the principal/username, group membership (if available), and the raw
 	// authentication data (if possible). This will allow other components in the pipeline to make decisions based on that data, such as routing based
 	// on tenancy as determined by the group membership, or passing through the authentication data to the next collector/backend.
 	// The context keys to be used are not defined yet.
-	Authenticate(ctx context.Context, headers map[string][]string) (context.Context, error)
+	Authenticate(ctx context.Context, requestMap map[string][]string) (context.Context, error)
```
Comment (jpkrohling) [inline config/configauth/serverauth.go:40]:
```suggestion
	Authenticate(ctx context.Context, metadata map[string][]string) (context.Context, error)
```

File: config/configauth/default_serverauthenticator.go [hunk 4]
```diff
@@ -57,7 +57,7 @@ func WithShutdown(shutdownFunc component.ShutdownFunc) Option {
 // NewServerAuthenticator returns a ServerAuthenticator configured with the provided options.
 func NewServerAuthenticator(options ...Option) ServerAuthenticator {
 	bc := &defaultServerAuthenticator{
-		AuthenticateFunc: func(ctx context.Context, headers map[string][]string) (context.Context, error) { return ctx, nil },
+		AuthenticateFunc: func(ctx context.Context, requestMap map[string][]string) (context.Context, error) { return ctx, nil },
```
[the remark this answers is not in the document]
Comment (neelayu) [reply config/configauth/default_serverauthenticator.go]: Yes. Good point. Renamed to `metadata`

File: config/confighttp/confighttp.go [hunk 5]
```diff
@@ -355,3 +356,30 @@ func maxRequestBodySizeInterceptor(next http.Handler, maxRecvSize int64) http.Ha
 		next.ServeHTTP(w, r)
 	})
 }
+
+func getParamsFromRequest(r *http.Request, propagate *configauth.PropagatePolicy) map[string][]string {
+	params := make(map[string][]string)
```
[the remark this answers is not in the document]
Comment (neelayu) [reply config/confighttp/confighttp.go]: Done

File: config/confighttp/confighttp_test.go [hunk 6]
```diff
@@ -863,7 +868,14 @@ func TestServerAuth(t *testing.T) {
 	require.NoError(t, err)
 
 	// test
-	srv.Handler.ServeHTTP(&httptest.ResponseRecorder{}, httptest.NewRequest("GET", "/", nil))
+	x := httptest.NewRequest("GET", "/", nil)
+	x.Header["x-test-header"] = []string{"test-value"}
+	q := x.URL.Query()
+	q.Add("query-test", "query-value1")
+	q.Add("query-test", "query-value2")
```
[the remark this answers is not in the document]
Comment (neelayu) [reply config/confighttp/confighttp_test.go]: Done

File: config/confighttp/confighttp_test.go [hunk 7]
```diff
@@ -911,6 +923,145 @@ func TestFailedServerAuth(t *testing.T) {
...
+				"query-test-3":    {"query-value-3"},
+			},
+		},
+		{
+			name: "propagate headers and query from policy",
+			args: args{
+				r: &http.Request{
+					Header: map[string][]string{
+						"x-test-header-1": {"test-value-1"},
+						"x-test-header-2": {"test-value-1"},
```
[the remark this answers is not in the document]
Comment (neelayu) [reply config/confighttp/confighttp_test.go]: Overlooked this. Thanks

**Comment by orloffv** [comment]
@jpkrohling codecov it's real problem? or if PR ready you can merge?

**Comment by jpkrohling** [comment]
I would like to try a more generic approach before approving this PR here.

**Comment by jpkrohling** [comment]
Sorry for taking so long to give further feedback, but I still think that we should first evaluate @bogdandrutu's suggestion as mentioned earlier:

> I believe @bogdandrutu mentioned that a more generic solution would be worth having. In that, components would register which values they need to have access to.

If that's not feasible, this PR would certainly come into consideration.

**Pull request closed without merging**